### ISSUE TITLE
Enhance toHaveValue's error message

### ISSUE THREAD

**Issue by mfelmy** [testing-library/jest-dom#218]
### Describe the feature you'd like:

I often forget/do not consider the type I've set on my input while making assertions on it. When I mistakenly expect the wrong type using the `toHaveValue` matcher, my test will -- obviously -- fail with, e.g.
```console
Expected the element to have value:
  8
Received:
  8
```
To already have an idea of what throws the error when looking at the console, the message could be enhanced with the values types
```console
Expected the element to have value:
  8 (number)
Received:
  8 (string)
```

### Suggested implementation:

When the expected and received value have loose equality, type information could be added to the output.

```js
  let expectedTypedValue = expectedValue;
  let receivedTypedValue = receivedValue;
  if (expectedValue == receivedValue)  {
    expectedTypedValue = `${expectedValue} (${typeof expectedValue})`;
    receivedTypedValue = `${receivedValue} (${typeof receivedValue})`;
  }
```

### Describe alternatives you've considered:

I did not consider any alternative solution since the expected changes are so small.

### Teachability, Documentation, Adoption, Migration Strategy:

No need for the user to adopt the way the library is used.

As many developers suggest to get started with contributing to open-source by improving the tools/libraries you already use and if you consider this change a helpful improvement on the DX, I would be really happy to submit my first PR to an open-source project I discovered recently and already fell in love with.

Thank you for this great library ❤

**Comment by gnapse** [comment]
This is a great suggestion indeed.

Would be great if you contributed the change as well. Are you up to?

The implementation is simple as you suggest. It's a matter of modifying [this file](https://github.com/testing-library/jest-dom/blob/master/src/to-have-value.js#L35-L41), and in particular get the customized typed values into lines 38 and 40 in that fragment over there.

**Comment by mfelmy** [comment]
Awesome!

Let me watch http://kcd.im/pull-request and then I will contribute my solution 😃

**Comment by eps1lon** [comment]
Does `toHaveValue` check attributes or properties? If it checks attributes it should flat out throw a TypeError if it doesn't receive a string. For properties writing out the type would indeed be nice.

**Comment by gnapse** [comment]
It checks the property. It even does some nice things to it (like gathering all selected option values from a multi-select into an array, or converting to number the value of a `<input type="number" />`). Check it out [here](https://github.com/testing-library/jest-dom/blob/57f03e9996847bcfac7f164c1d0307877f240cf7/src/utils.js#L173-L186).

**Comment by mfelmy** [comment]
To me it looks like it checks the [`value` property](https://github.com/testing-library/jest-dom/blob/master/src/utils.js#L149-L186) and not the `attributes` .

**Comment by afontcu** [comment]
Solved in #219
